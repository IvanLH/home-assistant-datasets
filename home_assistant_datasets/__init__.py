"""
.. include:: ../README.md
"""

__all__ = [
    "tool",
    "data_model",
    "model_client",
    "secrets",
]
