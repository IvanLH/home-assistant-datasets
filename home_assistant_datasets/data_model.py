"""Common librarys for collecting model outputs for evaluation."""

from dataclasses import dataclass, field
from functools import lru_cache
from typing import Any
import pathlib

from slugify import slugify
import yaml

from . import yaml_loaders


MODEL_CONFIG_FILE = pathlib.Path("models.yaml")
MODEL_CONFIG_DIR = pathlib.Path("models")
DATASET_CARD_FILE = "dataset_card.yaml"
DATASET_CARD_FILES = sorted(
    list(pathlib.Path("datasets").glob(f"**/{DATASET_CARD_FILE}"))
)
DATASET_TASK_IGNORE_FILES = {
    "_fixtures.yaml",
    "_home.yaml",
    "solution.yaml",
    DATASET_CARD_FILE,
}
# Label used to keep around configuration but not use for latest metric calculations
ARCHIVED_LABEL = "archived"


@dataclass(kw_only=True)
class DatasetCard:
    """A description of a dataset."""

    name: str
    """The name of the dataset."""

    description: str
    """A detailed description of the dataset."""

    urls: list[str]
    """More information about the dataset."""

    count: int | None = None
    """Number of times to run each task by default."""

    version: str | None = None
    """Current version of the dataset."""

    path: str | None = field(default=None)
    """Path to use for reading the dataset files instead of the current directory."""

    config_entry_data: dict[str, Any] | None = field(default=None)
    """Additional config entry data to include in model entry configs for this dataset."""

    config_entry_options: dict[str, Any] | None = field(default=None)
    """Additional config entry options to include in model entry configs for this dataset."""


@dataclass
class EntryConfig:
    """The configuration entry for the model under evaluation."""

    domain: str
    """The domain of the entry."""

    config_entry_data: dict[str, Any] | None = None
    """The configuration entry data."""

    config_entry_options: dict[str, Any] | None = None
    """The configuration entry options."""

    version: int | None = None
    """The version number of the config entry."""


@dataclass
class ModelConfig:
    """The configuration for the conversation agent under evaluation."""

    model_id: str
    """Details about the model name being evaluated if the domain supports multiple models."""

    domain: str
    """The domain under evaluation."""

    description: str
    """A detailed description of the model tested."""

    urls: list[str] | None = None
    """A list of relevant urls for the model and its serving infrastructure."""

    config_entry_data: dict[str, Any] | None = None
    """The configuration entry data."""

    config_entry_options: dict[str, Any] | None = None
    """The configuration entry options."""

    version: int | None = None
    """The version number of the config entry."""

    rpm: int | None = None
    """Requests per minute allowed for this model."""

    categories: list[str] = field(default_factory=list)
    """Arbitrary labels about this model."""


@dataclass
class Models:
    """The configuration for the full set of models in the config file."""

    models: list[ModelConfig]
    """The list of models under configuration"""

    prerequisites: list[EntryConfig]
    """The prerequisites for the models under evaluation."""


@lru_cache
def read_models() -> Models:
    """Read models configuration file."""
    if MODEL_CONFIG_FILE.exists():
        model_config_content = MODEL_CONFIG_FILE.open()
        try:
            models: Models = yaml_loaders.yaml_decode(model_config_content, Models)
        except Exception as err:
            raise ValueError(f"Error while loading {MODEL_CONFIG_FILE}: {err}")
    else:
        models = Models(models=[], prerequisites=[])

    for model_file in sorted(list(MODEL_CONFIG_DIR.glob("**/*.yaml"))):
        try:
            model_config = yaml_loaders.yaml_decode(model_file.open(), ModelConfig)
        except Exception as err:
            raise ValueError(f"Error while loading {model_file}: {err}")
        if "archive" in str(model_file):
            model_config.categories.append(ARCHIVED_LABEL)
        models.models.append(model_config)

    return models


def read_model(model_id: str) -> ModelConfig:
    """Read a specific model from the config file."""
    models = read_models()
    for model in models.models:
        if model.model_id == model_id:
            return model

    raise ValueError(
        f"Model config file '{MODEL_CONFIG_FILE}' did not contain model_id: {model_id}"
    )


@lru_cache
def read_dataset_card(dataset_card: pathlib.Path) -> DatasetCard:
    """Read a single dastaset configuration file."""
    data = yaml.load(dataset_card.read_text(), Loader=yaml.Loader)
    return DatasetCard(**data)


def read_prerequisites() -> list[EntryConfig]:
    """Read the prerequisites for the models under evaluation."""
    return read_models().prerequisites


def read_dataset_cards() -> list[DatasetCard]:
    """Read dataset configuration file."""
    cards = []
    for dataset_card in DATASET_CARD_FILES:
        cards.append(read_dataset_card(dataset_card))
    return cards


def _make_slug(text: str) -> str:
    """Shorthand slugify command"""
    return slugify(text, separator="_")


def read_dataset_files(dataset_path: pathlib.Path) -> dict[str, pathlib.Path]:
    """Generate the list of dataset files for the given path."""

    dataset_card = read_dataset_card(dataset_path / DATASET_CARD_FILE)
    if dataset_card.path:
        # The dataset card points to files in another directory
        dataset_path = pathlib.Path(dataset_card.path)

    def record_id(filename: pathlib.Path) -> str:
        return _make_slug(str(filename.relative_to(dataset_path))[:-5])

    # Tests are parameterized by the files that contain device actions. Ignore
    # fixtures and load those separately below.
    dataset_files = {
        record_id(filename): filename
        for filename in dataset_path.glob("**/*.yaml")
        if filename.name not in DATASET_TASK_IGNORE_FILES
    }
    if not dataset_files:
        raise ValueError(
            f"Could not find any dataset files in path: {str(dataset_path)}"
        )

    return dataset_files
