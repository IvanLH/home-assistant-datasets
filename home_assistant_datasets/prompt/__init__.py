"""Library for working with lower level details of model prompts."""
