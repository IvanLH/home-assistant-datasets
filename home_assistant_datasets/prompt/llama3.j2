{% if system or tools %}<|start_header_id|>system<|end_header_id|>
{% if system %}

{{ system }}
{%- endif %}
{%- if tools %}

You are a helpful assistant with tool calling capabilities. When you receive a tool call response, use the output to format an answer to the orginal use question.
{%- endif %}
{%- endif %}<|eot_id|>
{%- for message in messages %}
{%- if message.role == "user" %}<|start_header_id|>user<|end_header_id|>
{%- if tools and loop.last %}

Given the following functions, please respond with a JSON for a function call with its proper arguments that best answers the given prompt.

Respond in the format {"name": function name, "parameters": dictionary of argument name and its value}. Do not use variables.

{{ tools }}
{%- endif %}

{{ message.content }}<|eot_id|>{% if loop.last %}<|start_header_id|>assistant<|end_header_id|>

{% endif %}
{%- elif message.role == "assistant" %}<|start_header_id|>assistant<|end_header_id|>
{%- if tool_calls %}

{% for function in tool_calls: %}{"name": "{{ function.Name }}", "parameters": {{ function.Arguments }}}{% endfor %}
{%- else %}

{{ message.content }}{% if not loop.last %}<|eot_id|>{% endif %}
{%- endif %}
{%- elif message.role == "tool" %}<|start_header_id|>ipython<|end_header_id|>

{{ message.content }}<|eot_id|>{% if loop.last %}<|start_header_id|>assistant<|end_header_id|>

{% endif %}
{%- endif %}
{%- endfor %}
