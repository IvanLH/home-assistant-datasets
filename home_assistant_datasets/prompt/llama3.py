"""Library for creating llama3 prompts."""

from dataclasses import dataclass
from typing import Any
import json
import pathlib
from collections.abc import Generator
from mashumaro.mixins.yaml import DataClassYAMLMixin

from datasets import IterableDataset
from jinja2 import Environment, PackageLoader, select_autoescape

TEMPLATE_FILE = "llama3.j2"


@dataclass
class ToolCall(DataClassYAMLMixin):
    name: str
    arguments: dict[str, Any]

    def json(self) -> str:
        """Serialize the ToolCall as a json string for the prompt."""
        return json.dumps({"name": self.name, "parameters": self.arguments})


@dataclass
class Tool(DataClassYAMLMixin):
    name: str
    description: str
    parameters: dict[str, Any]

    def json(self) -> str:
        """Serialize the Tool as a json string for the prompt."""
        return json.dumps(
            {
                "type": "function",
                "function": {
                    "name": self.name,
                    "description": self.description,
                    "parameters": self.parameters,
                },
            },
            indent=4,
        )


@dataclass
class Message(DataClassYAMLMixin):
    role: str
    content: str | None = None
    tool_calls: list[ToolCall] | None = None


@dataclass
class ConversationRecord(DataClassYAMLMixin):
    instructions: str
    tools: list[Tool] | None
    input: str
    output: str | None
    tool_calls: list[ToolCall] | None = None


def build_prompt(messages: list[Message], tools: list[Tool] | None = None) -> str:
    """Build a llama3.1 prompt from the list of messages."""

    env = Environment(
        loader=PackageLoader("home_assistant_datasets", "prompt"),
        autoescape=select_autoescape(),
    )
    template = env.get_template(TEMPLATE_FILE)

    system: str | None = next(
        iter([message.content for message in messages if message.role == "system"]),
        None,
    )
    tools_json: str | None = (
        "\n".join([tool.json() for tool in tools]) if tools else None
    )
    return template.render(
        system=system or "",
        tools=tools_json,
        messages=messages,
    )


def build_prompt_record(record: ConversationRecord) -> str:
    """Build a llama3.1 prompt from a training record."""
    if not record.output and not record.tool_calls:
        raise ValueError(
            "Could not find assistant output to train, no output or tool_calls"
        )
    messages: list[Message] = [
        Message(role="system", content=record.instructions),
        Message(role="user", content=record.input),
        Message(role="assistant", content=record.output, tool_calls=record.tool_calls),
    ]
    return build_prompt(messages, record.tools)


def llama3_dataset_from_conversations(dataset_dir: pathlib.Path) -> IterableDataset:
    """Create a huggingface dataset from a training conversation record."""

    def func() -> Generator[dict[str, str], None, None]:
        for filename in dataset_dir.glob("*.yaml"):
            record = ConversationRecord.from_yaml(filename.read_text())
            yield {"text": build_prompt_record(record)}

    return IterableDataset.from_generator(func)
