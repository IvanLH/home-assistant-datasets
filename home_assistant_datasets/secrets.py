"""Library for loading secrets from secrets.yaml file.

This will read secrets from a `secrets.yaml` file or whatever filename
is referenced by the environment variable `SECRETS_FILE`.

The secrets file is a yaml key/value pair like this:

```yaml
---
google_api_key: XXXXXXXXXXXXXXXXXXXX
openai_api_key: sk-XXXXXXXXXXXXXXXXXXXX
vicuna_convesation_base_url: http://llama-cublas.llama:8000/v1
```

A secrets file is ignored by `.gitignore` allowing you to check in the code
for most of your git repo but leaving your secrets local.
"""

import yaml
import os

DEFAULT_SECRETS_FILE = "secrets.yaml"

def get_secret(secret_name: str) -> str:
    """Get secret with the specified name."""
    secrets_file = os.environ.get("SECRETS_FILE", DEFAULT_SECRETS_FILE)
    secrets = yaml.load(open(secrets_file), Loader=yaml.Loader)
    return secrets[secret_name]
