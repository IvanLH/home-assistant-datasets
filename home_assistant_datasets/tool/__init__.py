"""Home Assistant Datasets command line tools.

```
usage: home-assistant-datasets [-h] [--debug] {assist,leaderboard} ...

Home Assistant Datasets Utility

positional arguments:
  {assist,leaderboard}  Action

options:
  -h, --help            show this help message and exit
  --debug               Enable log output
```
"""

__all__ = [
    "assist",
    "leaderboard"
]
