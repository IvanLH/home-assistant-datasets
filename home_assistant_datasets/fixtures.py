"""Test fixtures for evaluations."""

from collections.abc import AsyncGenerator, Generator
import logging
import os
import pathlib
from typing import Any, TextIO, cast
from unittest.mock import patch, mock_open

import pytest
import pytest_socket
import yaml

from homeassistant.core import HomeAssistant
from homeassistant.config_entries import ConfigEntryState
from homeassistant.setup import async_setup_component

from custom_components import synthetic_home  # noqa: F401

# TODO(#12): Support loading from the custom component or core development environment
from pytest_homeassistant_custom_component.common import MockConfigEntry

from home_assistant_datasets.data_model import ModelConfig

from . import data_model


_LOGGER = logging.getLogger(__name__)


@pytest.fixture(autouse=True)
def mock_allow_sockets(socket_enabled: Any) -> None:
    """Remove all restrictions talking to sockets.

    This is needed when talking to external data sources
    like LLMs.
    """
    pytest_socket.pytest_runtest_teardown()
    pass


@pytest.fixture(autouse=True)
def auto_enable_custom_integrations(
    enable_custom_integrations: None,
) -> Generator[None, None, None]:
    """Enable custom integration."""
    _ = enable_custom_integrations  # unused
    yield


@pytest.fixture(autouse=True)
async def mock_default_components(hass: HomeAssistant) -> None:
    """Fixture to setup required default components."""
    assert await async_setup_component(hass, "homeassistant", {})
    assert await async_setup_component(hass, "conversation", {})


@pytest.fixture(name="synthetic_home_config")
def mock_synthetic_home_config() -> str | None:
    """Fixture to load the synthetic home config."""
    return None


@pytest.fixture(name="synthetic_home_yaml")
def mock_synthetic_home_content(synthetic_home_config: str | None) -> str | None:
    """Mock out the yaml config file contents."""
    if synthetic_home_config is None:
        return None
    with pathlib.Path(synthetic_home_config).absolute().open("r") as f:
        return f.read()


@pytest.fixture(autouse=True, name="synthetic_home_config_entry")
async def mock_synthetic_home(
    hass: HomeAssistant, synthetic_home_yaml: str | None
) -> AsyncGenerator[MockConfigEntry | None, None]:
    """Fixture for mock configuration entry."""
    if synthetic_home_yaml is None:
        yield None
        return

    config_entry = MockConfigEntry(
        domain="synthetic_home", data={"config_filename": "ignored"}
    )
    config_entry.add_to_hass(hass)

    with (
        patch(
            "synthetic_home.inventory.read_config_content",
            mock_open(read_data=synthetic_home_yaml),
        ),
        # Performance improvements during evaluation
        patch("custom_components.synthetic_home.cover.COVER_INSTANT", True),
    ):
        await hass.config_entries.async_setup(config_entry.entry_id)
        assert config_entry.state == ConfigEntryState.LOADED
        yield config_entry


@pytest.fixture(name="system_prompt")
async def system_prompt_fixture() -> None:
    """Fixture to provide the system prompt or None to use the default."""
    return None


@pytest.fixture
async def model_config(model_id: str) -> ModelConfig:
    """Fixture to read the model config yaml."""
    try:
        return data_model.read_model(model_id)
    except Exception as err:
        pytest.exit(str(err))


@pytest.fixture(name="conversation_agent_config_entry")
async def mock_conversation_agent_config_entry(
    hass: HomeAssistant, model_config: ModelConfig, system_prompt: str | None
) -> MockConfigEntry | None:
    if model_config.domain == "homeassistant":
        return None
    options = {}
    if model_config.config_entry_options:
        options.update(model_config.config_entry_options)
    if system_prompt:
        options["prompt"] = system_prompt
    config_entry = MockConfigEntry(
        domain=model_config.domain,
        data=model_config.config_entry_data,
        options=options,
        version=model_config.version or 1,
    )
    config_entry.add_to_hass(hass)
    await hass.config_entries.async_setup(config_entry.entry_id)
    assert config_entry.state == ConfigEntryState.LOADED
    return config_entry


class ConversationAgent:
    """A client library for a conversation agent service call."""

    def __init__(self, agent_id: str) -> None:
        """Initialize the agent."""
        self._agent_id = agent_id

    async def async_process(self, hass: HomeAssistant, text: str) -> str:
        """Process a text input and return the response."""
        service_response = await hass.services.async_call(
            "conversation",
            "process",
            {"agent_id": self._agent_id, "text": text},
            blocking=True,
            return_response=True,
        )
        assert service_response
        response = service_response["response"]
        return str(response["speech"]["plain"]["speech"])  # type: ignore[call-overload, index]


@pytest.fixture(name="conversation_agent_id")
async def mock_conversation_agent_id(
    model_config: ModelConfig,
    conversation_agent_config_entry: MockConfigEntry,
) -> str:
    """Return the id for the conversation agent under test."""
    if model_config.domain == "homeassistant":
        return "conversation.home_assistant"
    return "conversation.mock_title"  # conversation_agent_config_entry.entry_id)


@pytest.fixture(name="agent")
async def mock_agent(conversation_agent_id: str) -> ConversationAgent:
    """Create the conversation agent client id."""
    return ConversationAgent(conversation_agent_id)


class EvalRecordWriter:
    """Writes records to the eval output."""

    def __init__(self, filename: pathlib.Path):
        """Initialize EvalRecordWriter."""
        os.makedirs(filename.parent, exist_ok=True)
        self._eval_output = filename
        self._fd: TextIO | None = None

    def open(self) -> None:
        """Initialize the record writer."""
        self._fd = open(self._eval_output, "w")
        self._records = 0

    def write(self, record: dict[str, Any]) -> None:
        """Write an eval record."""
        assert self._fd
        self._fd.write(yaml.dump(record, sort_keys=False, explicit_start=True))
        self._fd.flush()
        self._records += 1

    def close(self) -> None:
        """Close the evaluation record."""
        _LOGGER.info(
            "Wrote %s summariies to %s",
            self._records,
            self._eval_output,
        )

        if self._fd is not None:
            self._fd.close()
            self._fd = None


@pytest.fixture(name="eval_record_writer")
def eval_record_writer_fixture(
    hass: HomeAssistant,
    eval_output_file: pathlib.Path,
) -> Generator[EvalRecordWriter, None, None]:
    """Fixture that prepares the eval output writer."""
    writer = EvalRecordWriter(eval_output_file)
    writer.open()
    yield writer
    writer.close()


@pytest.fixture(autouse=True)
def validate_entities(
    hass: HomeAssistant,
    synthetic_home_config_entry: MockConfigEntry,
    synthetic_home_yaml: str,
) -> None:
    """Fixture to verify that all entities are property created by synthetic home to avoid misconfiguration."""
    inventory = yaml.load(synthetic_home_yaml, Loader=yaml.Loader)
    assert inventory
    if not (entities := inventory.get("entities")):
        raise ValueError(
            f"No entities were specified in the inventory file: {inventory.keys()}"
        )
    for entity in entities:
        entity_id = entity["id"]
        state = hass.states.get(entity_id)
        assert state, f"Entity id not created {entity_id}"
        assert state.state != "unavailable"
        assert state.state not in (
            "unavailable",
            "unknown",
        ), f"Entity id has unavailable state {entity_id}: {state.state}"
