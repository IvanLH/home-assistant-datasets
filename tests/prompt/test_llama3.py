"""Tests for llama3 prompt building."""

from syrupy import SnapshotAssertion
import pathlib

from home_assistant_datasets.prompt import llama3


TOOLS = [
    llama3.Tool(
        name="get_current_conditions",
        description="Get the current weather conditions for a specific location",
        parameters={"location": {"type": "string"}},
    )
]
TOOL_CALLS = [
    llama3.ToolCall(
        name="get_current_conditions",
        arguments={"location": "San Francisco, CA"},
    )
]


def test_empty_messages(snapshot: SnapshotAssertion) -> None:
    """Test an empty list of messages."""

    assert llama3.build_prompt(messages=[]) == snapshot


def test_system(snapshot: SnapshotAssertion) -> None:
    """Test a single system message."""

    prompt = llama3.build_prompt(
        messages=[llama3.Message(role="system", content="You are a helpful assistant.")]
    )
    assert prompt == snapshot


def test_conversation(snapshot: SnapshotAssertion) -> None:
    """Test a conversation."""

    prompt = llama3.build_prompt(
        messages=[
            llama3.Message(
                role="system",
                content="You are a helpful assistant.",
            ),
            llama3.Message(role="user", content="What is the capital of France?"),
            llama3.Message(role="assistant", content="The capital of France is Paris."),
        ]
    )
    assert prompt == snapshot


def test_system_tools(snapshot: SnapshotAssertion) -> None:
    """Test a system message with tools."""

    prompt = llama3.build_prompt(
        messages=[
            llama3.Message(
                role="system",
                content="You are a helpful assistant.",
            ),
            llama3.Message(
                role="user",
                content="What is the weather like today in SF?",
            ),
        ],
        tools=TOOLS,
    )
    assert prompt == snapshot


def test_tool_call(snapshot: SnapshotAssertion) -> None:
    """Test a tool call with output."""

    prompt = llama3.build_prompt(
        messages=[
            llama3.Message(
                role="system",
                content="You are a helpful assistant.",
            ),
            llama3.Message(
                role="user",
                content="What is the weather like today in SF?",
            ),
            llama3.Message(
                role="assistant",
                tool_calls=TOOL_CALLS,
            ),
            llama3.Message(
                role="tool",
                content='{"output": "Clouds giving way to sun Hi: 76째 Tonight: Mainly clear early, then areas of low clouds forming Lo: 56째"}',
            ),
        ],
        tools=TOOLS,
    )
    assert prompt == snapshot


def test_tool_call_conversation(snapshot: SnapshotAssertion) -> None:
    """Test a tool call conversation."""

    prompt = llama3.build_prompt(
        messages=[
            llama3.Message(
                role="system",
                content="You are a helpful assistant.",
            ),
            llama3.Message(
                role="user",
                content="What is the weather like today in SF?",
            ),
            llama3.Message(
                role="assistant",
                tool_calls=TOOL_CALLS,
            ),
            llama3.Message(
                role="tool",
                content='{"output": "Clouds giving way to sun Hi: 76째 Tonight: Mainly clear early, then areas of low clouds forming Lo: 56째"}',
            ),
            llama3.Message(
                role="assistant",
                content="The weather in SF looks cloudy today.",
            ),
        ],
        tools=TOOLS,
    )
    assert prompt == snapshot


def test_converation_record(snapshot: SnapshotAssertion) -> None:
    """Test a training record."""

    message = llama3.ConversationRecord.from_dict(
        {
            "instructions": "You are a helpful assistant.",
            "tools": [tool.to_dict() for tool in TOOLS],
            "input": "What is the weather like today in SF?",
            "output": "I will call a weather tool",
            "tool_calls": [call.to_dict() for call in TOOL_CALLS],
        }
    )
    prompt = llama3.build_prompt_record(message)
    assert prompt == snapshot


def test_converation_record_output(snapshot: SnapshotAssertion) -> None:
    """Test a training record."""

    message = llama3.ConversationRecord.from_dict(
        {
            "instructions": "You are a helpful assistant.",
            "tools": [tool.to_dict() for tool in TOOLS],
            "input": "What is the weather like today in SF?",
            "output": "I can't help you with that.",
            "tool_calls": None,
        }
    )
    prompt = llama3.build_prompt_record(message)
    assert prompt == snapshot


def test_llama3_dataset_from_conversations(snapshot: SnapshotAssertion) -> None:
    """Test generating a huggingface dataset from conversation records."""

    ds = llama3.llama3_dataset_from_conversations(
        pathlib.Path("datasets/device-actions-v2-collect/train")
    )
    record = next(iter(ds))
    assert "text" in record
    assert record["text"] == snapshot
