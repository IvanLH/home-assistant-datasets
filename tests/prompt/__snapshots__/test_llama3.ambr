# serializer version: 1
# name: test_converation_record
  '''
  <|start_header_id|>system<|end_header_id|>


  You are a helpful assistant.

  You are a helpful assistant with tool calling capabilities. When you receive a tool call response, use the output to format an answer to the orginal use question.<|eot_id|><|start_header_id|>user<|end_header_id|>

  Given the following functions, please respond with a JSON for a function call with its proper arguments that best answers the given prompt.

  Respond in the format {"name": function name, "parameters": dictionary of argument name and its value}. Do not use variables.

  {
      "type": "function",
      "function": {
          "name": "get_current_conditions",
          "description": "Get the current weather conditions for a specific location",
          "parameters": {
              "location": {
                  "type": "string"
              }
          }
      }
  }

  What is the weather like today in SF?<|eot_id|><|start_header_id|>assistant<|end_header_id|>

  I will call a weather tool
  '''
# ---
# name: test_conversation
  '''
  <|start_header_id|>system<|end_header_id|>


  You are a helpful assistant.<|eot_id|><|start_header_id|>user<|end_header_id|>

  What is the capital of France?<|eot_id|><|start_header_id|>assistant<|end_header_id|>

  The capital of France is Paris.
  '''
# ---
# name: test_empty_messages
  '<|eot_id|>'
# ---
# name: test_system
  '''
  <|start_header_id|>system<|end_header_id|>


  You are a helpful assistant.<|eot_id|>
  '''
# ---
# name: test_system_tools
  '''
  <|start_header_id|>system<|end_header_id|>


  You are a helpful assistant.

  You are a helpful assistant with tool calling capabilities. When you receive a tool call response, use the output to format an answer to the orginal use question.<|eot_id|><|start_header_id|>user<|end_header_id|>

  Given the following functions, please respond with a JSON for a function call with its proper arguments that best answers the given prompt.

  Respond in the format {"name": function name, "parameters": dictionary of argument name and its value}. Do not use variables.

  {
      "type": "function",
      "function": {
          "name": "get_current_conditions",
          "description": "Get the current weather conditions for a specific location",
          "parameters": {
              "location": {
                  "type": "string"
              }
          }
      }
  }

  What is the weather like today in SF?<|eot_id|><|start_header_id|>assistant<|end_header_id|>


  '''
# ---
# name: test_tool_call
  '''
  <|start_header_id|>system<|end_header_id|>


  You are a helpful assistant.

  You are a helpful assistant with tool calling capabilities. When you receive a tool call response, use the output to format an answer to the orginal use question.<|eot_id|><|start_header_id|>user<|end_header_id|>

  Given the following functions, please respond with a JSON for a function call with its proper arguments that best answers the given prompt.

  Respond in the format {"name": function name, "parameters": dictionary of argument name and its value}. Do not use variables.

  {
      "type": "function",
      "function": {
          "name": "get_current_conditions",
          "description": "Get the current weather conditions for a specific location",
          "parameters": {
              "location": {
                  "type": "string"
              }
          }
      }
  }

  What is the weather like today in SF?<|eot_id|><|start_header_id|>assistant<|end_header_id|>

  <|eot_id|><|start_header_id|>ipython<|end_header_id|>

  {"output": "Clouds giving way to sun Hi: 76째 Tonight: Mainly clear early, then areas of low clouds forming Lo: 56째"}<|eot_id|><|start_header_id|>assistant<|end_header_id|>


  '''
# ---
# name: test_tool_call_conversation
  '''
  <|start_header_id|>system<|end_header_id|>


  You are a helpful assistant.

  You are a helpful assistant with tool calling capabilities. When you receive a tool call response, use the output to format an answer to the orginal use question.<|eot_id|><|start_header_id|>user<|end_header_id|>

  Given the following functions, please respond with a JSON for a function call with its proper arguments that best answers the given prompt.

  Respond in the format {"name": function name, "parameters": dictionary of argument name and its value}. Do not use variables.

  {
      "type": "function",
      "function": {
          "name": "get_current_conditions",
          "description": "Get the current weather conditions for a specific location",
          "parameters": {
              "location": {
                  "type": "string"
              }
          }
      }
  }

  What is the weather like today in SF?<|eot_id|><|start_header_id|>assistant<|end_header_id|>

  <|eot_id|><|start_header_id|>ipython<|end_header_id|>

  {"output": "Clouds giving way to sun Hi: 76째 Tonight: Mainly clear early, then areas of low clouds forming Lo: 56째"}<|eot_id|><|start_header_id|>assistant<|end_header_id|>

  The weather in SF looks cloudy today.
  '''
# ---
